WordPress-iOS-Shared/Core/OperationPools.swift

```diff
@@ -0,0 +1,286 @@
+import UIKit
+
+/// OperationPool is an abstraction to coalesce deterministic operations.
+///
+/// When you have an _expensive_ operation like a network request or resizing a
+/// large image, it is common to resort to caching. But caching only avoids
+/// extra operations once the first one has finished.
+///
+/// OperationPool takes care of coalescing multiple transformations for the same
+/// resource, executing only once, and notifying of completion to all the
+/// requesters.
+///
+/// Each operation transforms a `Key` into a `Value`. This could be a URL to a
+/// data object, an image to a smaller image, or anything else. In any case, the
+/// same `Key` should always produce the same `Value`.
+///
+/// Defining custom operation pools
+/// -------------------------------
+///
+/// To define your own custom operations you can subclass OperationPool. All of
+/// the defined methods are `final` so you are not supposed to override
+/// anything.
+///
+/// Instead, define your own convenience initializer and call `init(_:)` with
+/// your custom `start` function.
+///
+/// An alternative is to use a factory function instead:
+///
+///     func RequestPool(session: NSURLSession) -> OperationPool<NSURLRequest, NSData, NSError> {
+///         return OperationPool { request, completion in
+///             let task = session.dataTaskWithRequest(request) {
+///                 data, response, error in
+///     
+///                 let result = Result(
+///                     value: data,
+///                     error: error,
+///                     fallbackError: NetworkError.noDataAndNoError.nserror
+///                 )
+///                 completion(result)
+///             }
+///             task.resume()
+///         }
+///     }
+/// 
+/// Subclassing operation pools is only meant as a sort of typealias plus a
+/// constuctor. If you need something more complicated, the recommendation is to
+/// create a different type that just holds an instance of an operation pool.
+///
+/// Threading considerations
+/// ------------------------
+///
+/// OperationPool is thread safe and can be called from any thread. Optionally,
+/// it can be initialized with a queue in which the completion blocks will be
+/// called. By default it uses a global low priority concurrent queue.
+///
+/// When an operation finishes, all the completion blocks are queued
+/// asynchronously at the same time.
+///
+/// Note that there's a delay between the time the `start` function calls its
+/// `completion` function and the completion blocks are called. If you want to
+/// add some caching of results, the recommended way is to handle storing the
+/// cached values from the `start` function to avoid duplicate operations.
+///
+public class OperationPool<Key: Hashable, Value, Error: ErrorType> {
```

Comment:
You almost scared me for a second 😉 
Yes, it's possible that two keys would have the same hash (although unlikely). It's also not a problem.

The `Key` is `Hashable` because we're using a dictionary internally, and it's a requirement. `Hashable` also requires `Equatable`, so we can have a way to check if those objects that have the same value are actually equal. `Hashable` is an optimization so a `Dictionary` can do fast lookup, but it still verifies equality.

Just to double check I ran this on a playground:

```swift
struct FakeHash: Hashable {
    let value: Int
    var hashValue: Int {
        return 1
    }
}

func == (lhs: FakeHash, rhs: FakeHash) -> Bool {
    return lhs.value == rhs.value
}

let item1 = FakeHash(value: 1)
let item2 = FakeHash(value: 2)
var dict = [FakeHash: String]()
dict[item1] = "1"
dict[item2] = "2"

item1 == item2
// false
item1.hashValue == item2.hashValue
// true
dict.description
// [FakeHash(value: 1): "1", FakeHash(value: 2): "2"]
```